### DISCUSSION TITLE
Custom HTTP methods

### DISCUSSION THREAD

**Discussion by connorjburton** [mswjs/msw#1196] [Q&A] [answered]
### Scope

Adds a new behavior

### Compatibility

- [ ] This is a breaking change

### Feature description

Some services that I have been mocking, such as [Vault](https://www.hashicorp.com/products/vault), use custom HTTP methods like `LIST`.

I think it would useful to be able to register custom HTTP methods, currently I have to use `.all` (which is undocumented) and use `req.method === 'LIST'`.

The API could be something as simple as `server.registerMethod('LIST')`

**Comment by kettanaito** [comment, marked as answer]
Hey, @connorjburton.

You can create your own `rest.list()` request handler like so:

```ts
// my-rest.js
import { rest as defaultRest, RestHandler } from 'msw'

function list(path, resolver) {
  return new RestHandler('list', path, resolver)
}

export const rest = {
  ...defaultRest,
  list,
}
```

And then use it in your handlers:

```js
import { rest } from './my-rest'

rest.list('/resource', resolver)
```

**Comment by connorjburton** [comment]
Thank you!

Is this documented anywhere? I couldn't find anything like this on the docs. Maybe it's worth adding as a recipe?

**Comment by kettanaito** [comment]
I'm writing a recipe on this but it's not ready just yet. The topic is quite complex as you can build _a lot_ of customization using our existing API. I'm also working on the next version of the documentation, and this recipe will likely be included there.
